### PR TITLE
Support register/update/list Smart Contract

### PR DESCRIPTION
**Pull request by jianlunz-cb**
### What changed? Why?
Support register/update/list Smart Contract 

Test:
```
> let s = await SmartContract.register("base-sepolia", "0x0000000071727de22e5e9d8baf0edac6f37da052", [{constant: true,name: "myField",outputs: [{name: "",type: "uint256"}],type: "function"}], "Test")
undefined
> console.log(s)
SmartContract {
  model: {
    abi: '[{"constant":true,"name":"myField","outputs":[{"name":"","type":"uint256"}],"type":"function"}]',
    contract_address: '0x0000000071727De22E5E9D8BAF0eDac6F37da052',
    contract_name: 'Test',
    is_external: true,
    network_id: 'base-sepolia',
    smart_contract_id: '7c3437ea-9a74-42a6-8778-ee83b9bbd4bb',
    type: 'custom'
  }
}
undefined
> await s.update([{constant: false, name: "myField2",outputs: [{name: "",type: "uint256"}],type: "function"}], "Test2")
SmartContract {
  model: {
    abi: '[{"constant":false,"name":"myField2","outputs":[{"name":"","type":"uint256"}],"type":"function"}]',
    contract_address: '0x0000000071727De22E5E9D8BAF0eDac6F37da052',
    contract_name: 'Test2',
    is_external: true,
    network_id: 'base-sepolia',
    smart_contract_id: '7c3437ea-9a74-42a6-8778-ee83b9bbd4bb',
    type: 'custom'
  }
}
> let l = await SmartContract.list()
undefined
> console.log(l.data)
[
  SmartContract {
    model: {
      abi: '[{"constant":false,"name":"myField2","outputs":[{"name":"","type":"uint256"}],"type":"function"}]',
      contract_address: '0x0000000071727de22e5e9d8baf0edac6f37da052',
      contract_name: 'Test2',
      is_external: false,
      network_id: 'base-sepolia',
      smart_contract_id: '',
      type: 'custom'
    }
  },
...
]
```

### CODE REVIEW

File: src/tests/smart_contract_test.ts [hunk 1]
```diff
@@ -55,6 +56,121 @@ describe("SmartContract", () => {
...
+      Coinbase.apiClients.smartContract!.registerSmartContract = jest
+        .fn()
+        .mockRejectedValue(new Error("Failed to register the smart contract"));
+      await expect(
+        SmartContract.register(networkId, contractAddress, testAllReadTypesABI, contractName),
+      ).rejects.toThrow("Failed to register the smart contract");
+    });
+  });
+
+  describe("SmartContract.update", () => {
```
Comment (alex-stone) [inline src/tests/smart_contract_test.ts:99]:
```suggestion
  describe(".update", () => {
```

File: src/tests/smart_contract_test.ts [hunk 2]
```diff
@@ -55,6 +56,121 @@ describe("SmartContract", () => {
     });
   });
 
+  describe("SmartContract.register", () => {
```
Comment (alex-stone) [inline src/tests/smart_contract_test.ts:59]:
```suggestion
  describe(".register", () => {
```

File: src/tests/smart_contract_test.ts [hunk 3]
```diff
@@ -55,6 +56,121 @@ describe("SmartContract", () => {
...
+    const networkId = erc20Model.network_id;
+    const contractName = erc20Model.contract_name;
+    const contractAddress = erc20Model.contract_address;
```
Comment (alex-stone) [inline src/tests/smart_contract_test.ts:62]: Is the erc20Model fixture an external contract? Probably worth making an external contract for this (We can fast-follow with this though)

**Review by alex-stone — APPROVED** [review]
Functionality looks reasonable.

I think we may need to do a pass on the pagination construct in NodeJS, but we can do that separately at the SDK layer.

Some test changes that we can follow up on to unblock the release

File: src/tests/smart_contract_test.ts (hunk 3 above)
Comment (jianlunz-cb) [reply src/tests/smart_contract_test.ts]: Sure, I've added a new external model

File: src/coinbase/smart_contract.ts [hunk 4]
```diff
@@ -100,6 +102,67 @@ export class SmartContract {
...
+   * @param networkId - The network ID.
+   * @param contractAddress - The contract address.
+   * @param abi - The ABI of the contract.
+   * @param contractName - The contract name.
+   * @returns The smart contract.
+   */
+  public static async register(
+    networkId: string,
+    contractAddress: string,
+    abi: object,
```
Comment (alex-stone) [inline src/coinbase/smart_contract.ts:117]: Does this enable this to be optional?

File: src/tests/smart_contract_test.ts [hunk 5]
```diff
@@ -17,6 +17,7 @@ import {
   VALID_SMART_CONTRACT_ERC1155_MODEL,
   ERC1155_URI,
   VALID_SMART_CONTRACT_EXTERNAL_MODEL,
+  testAllReadTypesABI, VALID_EXTERNAL_SMART_CONTRACT_ERC20_MODEL,
```
Comment (alex-stone) [inline src/tests/smart_contract_test.ts:20]:
```suggestion
  testAllReadTypesABI,
  VALID_EXTERNAL_SMART_CONTRACT_ERC20_MODEL,
```

File: src/tests/smart_contract_test.ts [hunk 6]
```diff
@@ -55,6 +58,118 @@ describe("SmartContract", () => {
...
+      Coinbase.apiClients.smartContract!.registerSmartContract = jest
+        .fn()
+        .mockRejectedValue(new Error("Failed to register the smart contract"));
+      await expect(
+        SmartContract.register(networkId, contractAddress, testAllReadTypesABI, contractName),
+      ).rejects.toThrow("Failed to register the smart contract");
+    });
+  });
+
+  describe(".update", () => {
```
Comment (alex-stone) [inline src/tests/smart_contract_test.ts:102]: Can we fast follow with tests for when name or ABI are omitted, since they are optional?

**Review by alex-stone — APPROVED** [review]

File: src/tests/smart_contract_test.ts [hunk 7]
```diff
@@ -55,6 +59,118 @@ describe("SmartContract", () => {
...
+          contract_name: updatedContract.contract_name,
+        },
+      );
+      expect(smartContract).toBeInstanceOf(SmartContract);
+      expect(smartContract.getContractAddress()).toBe(contractAddress);
+      expect(smartContract.getAbi()).toEqual(updatedAbiJson);
+      expect(smartContract.getContractName()).toEqual(updatedContract.contract_name);
+    });
+
+    it("should throw an error if creation fails", async () => {
```
Comment (alex-stone) [inline src/tests/smart_contract_test.ts:136]:
```suggestion
    it("should throw an error if updating fails", async () => {
```

File: src/tests/smart_contract_test.ts (hunk 2 above)
Comment (alex-stone) [reply src/tests/smart_contract_test.ts:62]: Curious why we reverted this change in the latest commit?

**Review by alex-stone — APPROVED** [review]

File: src/tests/smart_contract_test.ts (hunk 6 above)
Comment (jianlunz-cb) [reply src/tests/smart_contract_test.ts:102]: Yep, I'll have a follow up diff to address it!

**Pull request merged**